### PR TITLE
Tests/chat server monadic tests

### PR DESCRIPTION
**Pull request by Kalior**
Tests for `Server.Chat` module

Part of #89

### CODE REVIEW

**Comment by Tejpbit** [comment]
Lots of hlint suggestions. Fix those and I'll come around for another review.

**Comment by Kalior** [comment]
Fixed "lots" of hlint suggestions

File: test/Server/ChatTest.hs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+-- stack ghci: :l test/Server/ChatTest.hs test/ArbitraryLobbyTypes.hs
+
+-- |Property that tests that after joining a chat the chat is created
+-- or the player has that chat in it's list
+prop_joinChat :: Int -> [String] -> [ClientEntry] -> Property
+prop_joinChat i chatNameList clientList = monadicIO $ do
+  pre $ not $ null clientList
+  pre $ not $ null chatNameList
+  let i' = abs $ mod i $ length clientList
```
Comment (Tejpbit) [inline test/Server/ChatTest.hs]: Use mod as infix operator.

What is `i'` ? The index of a random channel which we'll be joining later? Consider renaming variable, the same goes for `i''`

File: test/Server/ChatTest.hs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+import ArbitraryLobbyTypes ()
+import LobbyTypes
+
+
+-- stack ghci: :l test/Server/ChatTest.hs test/ArbitraryLobbyTypes.hs
+
+-- |Property that tests that after joining a chat the chat is created
+-- or the player has that chat in it's list
+prop_joinChat :: Int -> [String] -> [ClientEntry] -> Property
+prop_joinChat i chatNameList clientList = monadicIO $ do
```
Comment (Tejpbit) [inline test/Server/ChatTest.hs]: What is i? Not as self explanitory as chatNameList and clientList...

File: test/Server/ChatTest.hs [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+  run $ Server.Chat.joinChat mVarClients mVarChats sid chat
+
+  newClientList <- run $ readMVar mVarClients
+  newChatList <- run $ readMVar mVarChats
+  let newClient = newClientList !! i'
+  let clientChats = chats newClient
+
+  assert $
+    any ((chat == ). fst) clientChats &&
```
Comment (Tejpbit) [inline test/Server/ChatTest.hs]: I think `chat`elem`map fst clientChats` would be more intuitive.

File: test/Server/ChatTest.hs [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    any ((chat == ). fst) newChatList
+
+-- |Property for leaveChat, tests that the chat is no longer in the players
+-- list.
+prop_leaveChat :: Int -> [ClientEntry] -> Property
```
Comment (Tejpbit) [inline test/Server/ChatTest.hs]: Same general comments here. Unclear what `i`, `i'` and `i''` is.

File: test/Server/ChatTest.hs [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+  let sid = sessionID client
+
+  mVarClients <- run $ newMVar clientList
+  run $ Server.Chat.leaveChat mVarClients sid chatName
+
+  newClientList <- run $ readMVar mVarClients
+  let newClient = newClientList !! i'
+  let clientChats = chats newClient
+
+  assert $ all ((chatName /=) . fst) clientChats
```
Comment (Tejpbit) [inline test/Server/ChatTest.hs]: Could use chatName `notElem` map fst clientChats.

**Comment by Tejpbit** [comment]
Great tests. Some minor code style and naming comments which could be discussed.

**Pull request merged**